docs/rfcs/036-physical-replication.md

```diff
@@ -0,0 +1,263 @@
+queries. They work similarly to read replicas that follow the primary,
+but some things are simpler: there are no concerns about cache
+invalidation when the data changes on the primary, or worrying about
+transactions that are in-progress on the primary.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+- Control plane launches the replica
+- Replica Postgres instance connects to the safekeepers, to stream the WAL
+- The primary does not know about the standby, except for the hot standby feedback
+- The primary and replicas all connect to the same pageservers
+
+
+# Context
+
+Some useful things to know about hot standby and replicas in
+PostgreSQL.
+
+## PostgreSQL startup sequence
+
+"Running" and "start up" terms are little imprecise. PostgreSQL
+replica startup goes through several stages:
+
+1. First, the process is started up, and various initialization steps
+   are performed, like initializing shared memory. If you try to
+   connect to the server in this stage, you get an error: ERROR: the
+   database system is starting up. This stage happens very quickly, no
+
+2. Then the server reads the checpoint record from the WAL and starts
+   the WAL replay starting from the checkpoint. This works differently
+   in Neon: we start the WAL replay at the basebackup LSN, not from a
+   checkpoint! If you connect to the server in this state, you get an
+   error: ERROR: the database system is not yet accepting
+   connections. We proceed to the next stage, when the WAL replay sees
+   a running-xacts record. Or in Neon, the "CLOG scanning" mechanism
+   can allow us to move directly to next stage, with all the caveats
+   listed in this RFC.
+
+3. When the running-xacts information is established, the server
+   starts to accept connections normally.
+
+From PostgreSQL's point of view, the server is already running in
+stage 2, even though it's not accepting connections yet. Our
+`compute_ctl` does not consider it as running until stage 3. If the
+transition from stage 2 to 3 doesn't happen fast enough, the control
+plane will mark the start operation as failed.
+
+
+## Decisions, Issues
+
+### Cache invalidation in replica
+
+When a read replica follows the primary in PostgreSQL, it needs to
+stream all the WAL from the primary and apply all the records, to keep
+the local copy of the data consistent with the primary. In Neon, the
+replica can fetch the updated page versions from the pageserver, so
+it's not necessary to apply all the WAL. However, it needs to ensure
+that any pages that are currently in the Postgres buffer cache, or the
+Local File Cache, are either updated, or thrown away so that the next
+read of the page will fetch the latest version.
+
+We choose to apply the WAL records for pages that are already in the
+buffer cache, and skip records for other pages. Somewhat arbitrarily,
+we also apply records affecting catalog relations, fetching the old
```

Comment:
I think that there is one question which should be clarified here: should we apply FPI when we receive them at replica? I have created PR for it long timer ago: https://github.com/neondatabase/neon/pull/7875
@MMeent is against it, because "it pollutes the cache". I strongly disagree with it.
Why do we have replicas at all? Not for HA (as mostly in vanilla Postgres), but for load balancing of read-only queries. Certainly it is possible that working set at primary and standby are completely different. But IMHO it is unlikely. And throwing aways received FPI we just do wasteful job and double network traffic. 

Also, there is some specific of current Neon setup which makes the current behaviour especially inadequate:
we now have small shared buffers (128Mb) and larger LFC  (100GB). If target page is not found in shared buffers, it is also thrown from LFC.  So small shared buffers doesn't;t allow us to hold working set and make it in sync with primary. And deleting page from LFC makes performance of replica really awful.

Reply:
yeah, it's not clear our policy or what to apply or not is optimal. This just documents the status quo. (If we knew what the best approach is, we should document the desired end state, but I don't think we know that)